src/bci_build/package/obs_package.py

```diff
@@ -0,0 +1,217 @@
+together multiple base container images into a single package.
+
+"""
+
+import abc
+import asyncio
+import os.path
+import xml.etree.ElementTree as ET
+from dataclasses import dataclass
+from io import BytesIO
+from typing import TYPE_CHECKING
+from typing import Coroutine
+from typing import Sequence
+
+from bci_build.constants import BuildType
+from bci_build.os_version import OsVersion
+from bci_build.service import Service
+from bci_build.util import write_to_file
+
+if TYPE_CHECKING:
+    from bci_build.package import BaseContainerImage
+
+
+@dataclass(kw_only=True)
+class ObsPackageBase(abc.ABC):
+    """Abstract base class of the ObsPackage and the BaseContainerImage."""
+
+    #: The name of the package in the Build Service
+    package_name: str | None = None
+
+    #: The OS version to which this package belongs
+    os_version: OsVersion
+
+    #: Define whether this container image is built using docker or kiwi.
+    #: If not set, then the build type will default to docker from SP4 onwards.
+    build_recipe_type: BuildType | None = None
+
+    def __post_init__(self) -> None:
+        if self.build_recipe_type is None:
+            self.build_recipe_type = (
+                BuildType.KIWI if self.os_version == OsVersion.SP3 else BuildType.DOCKER
+            )
+
+    @property
+    @abc.abstractmethod
+    def uid(self) -> str:
+        """unique identifier of this package, either its name or ``$name-$tag_version``."""
+
+    @property
+    @abc.abstractmethod
+    def services(self) -> tuple[Service, ...]:
+        """The source services that are part of this package."""
+
+    @property
+    @abc.abstractmethod
+    def title(self) -> str:
+        """The title of this package."""
+
+    @property
+    @abc.abstractmethod
+    def description(self) -> str:
+        """The description of this package."""
+
+    @abc.abstractmethod
+    async def write_files_to_folder(
+        self, dest: str, *, with_service_file: bool = True
+    ) -> list[str]:
+        """Write all files belonging to this package into the directory
+        ``dest``.
+
+        If ``with_service_file`` is ``False``, then the :file:`_service` will
+        not be written to ``dest``.
+
+        """
+
+    @property
+    def _service_file_contents(self) -> str:
+        root = ET.Element("services")
+        for service in [
+            Service(name=f"{self.build_recipe_type}_label_helper"),
+            Service(name="kiwi_metainfo_helper"),
+        ] + list(self.services):
+            root.append(service.as_xml_element())
+
+        tree = ET.ElementTree(root)
+        ET.indent(tree)
+        io = BytesIO()
+        tree.write(io, encoding="utf-8")
+        io.seek(0)
+        return io.read().decode("utf-8")
+
+    async def _write_service_file(self, dest: str) -> list[str]:
+        await write_to_file(os.path.join(dest, "_service"), self._service_file_contents)
+        return ["_service"]
```

Comment:
that return value seems unused? since it should end up in the files list anyway..

Reply:
It is used here: https://github.com/SUSE/BCI-dockerfile-generator/blob/7e43203c16fa10c06faa7abea7eaa9e2acf676df/src/bci_build/package/obs_package.py#L187